internal-dns/src/dns_server.rs

```diff
@@ -67,21 +74,51 @@ async fn handle_req<'a, 'b, 'c>(
 
     println!("{:#?}", mr);
 
-    let rb = MessageResponseBuilder::from_message_request(&mr);
     let header = Header::response_from_request(mr.header());
+    let zone = LowerName::from(Name::from_str(&zone).unwrap());
+
+    // Ensure the query is for this zone, otherwise bail with servfail. This
+    // will cause resolvers to look to other DNS servers for this query.
+    let name = mr.query().name();
+    if !zone.zone_of(name) {
+        nack(&log, &mr, &socket, &header, &src).await;
+        return;
+    }
 
     let name = mr.query().original().name().clone();
     let key = name.to_string();
     let key = key.trim_end_matches('.');
 
+    let rb = MessageResponseBuilder::from_message_request(&mr);
+
     let bits = match db.get(key.as_bytes()) {
         Ok(Some(bits)) => bits,
-        Err(e) => {
-            error!(log, "db get: {}", e);
-            nack(&log, &mr, &socket, &header, &src).await;
+
+        // If no record is found bail with NXDOMAIN.
+        Ok(None) => {
+            let mresp = rb.error_msg(&header, ResponseCode::NXDomain);
+            let mut resp_data = Vec::new();
+            let mut enc = BinEncoder::new(&mut resp_data);
+            match mresp.destructive_emit(&mut enc) {
+                Ok(_) => {}
+                Err(e) => {
+                    error!(log, "NXDOMAIN destructive emit: {}", e);
+                    nack(&log, &mr, &socket, &header, &src).await;
+                    return;
+                }
+            }
+            match socket.send_to(&resp_data, &src).await {
+                Ok(_) => {}
+                Err(e) => {
+                    error!(log, "NXDOMAIN send: {}", e);
```

Comment:
Do we not need to `nack` here? I see we're doing it on most other error paths

Reply:
I also noticed it missing with the `deserialize record` "error!" call below, fwiw

Reply:
Good catch. There should be `nack` on the deserialize error. For the `socket.send_to` errors there's not much we can do as the `nack` also needs to call `socket.send_to` to get the nack to the client.